crates/circuit/src/symbol_expr.rs

```diff
@@ -28,16 +29,16 @@ pub const SYMEXPR_EPSILON: f64 = f64::EPSILON * 8.0;
 /// node types of expression tree
 #[derive(Debug, Clone)]
 pub enum SymbolExpr {
-    Symbol(Box<String>),
+    Symbol(Arc<String>),
     Value(Value),
     Unary {
         op: UnaryOp,
-        expr: Box<SymbolExpr>,
+        expr: Arc<SymbolExpr>,
     },
     Binary {
```

Comment:
At least for `Expr`, I put expression variants with multiple inner `Expr`s within a single smart pointer, e.g. in this case:

```rust
struct Binary {
    op: BinaryOp,
    lhs: SymbolExpr,
    rhs: SymbolExpr,
}

pub enum SymbolExpr {
    Symbol(Box<String>),
    Symbol(Arc<String>),
    Value(Value),
    Unary {
        op: UnaryOp,
        expr: Arc<SymbolExpr>,
    },
    Binary(Arc<Binary>),
}
```

I think maybe @doichanj had this as well originally. Perhaps that could help reduce fan-out.